From 5ed99bce77fa0575f69e96e0ce33a504c7edc63d Mon Sep 17 00:00:00 2001
From: Binyuan Lan <lby@rock-chips.com>
Date: Thu, 8 Jul 2021 21:45:30 +0800
Subject: [PATCH] ASoC: rockchip: rk817-codec: Solve pop problems

DIGEN_CLKE can only be controlled when the amplifier is turned off

Signed-off-by: Binyuan Lan <lby@rock-chips.com>
Change-Id: I5758cab8f20bd2ede008d1ad138cdadcaa55e39d
---
 sound/soc/codecs/rk817_codec.c | 15 ++++++++-------
 1 file changed, 8 insertions(+), 7 deletions(-)

diff --git a/sound/soc/codecs/rk817_codec.c b/sound/soc/codecs/rk817_codec.c
index 59db9a874324..1b5bc27fd66a 100644
--- a/sound/soc/codecs/rk817_codec.c
+++ b/sound/soc/codecs/rk817_codec.c
@@ -831,24 +831,25 @@ static int rk817_digital_mute(struct snd_soc_dai *dai, int mute)
 	struct rk817_codec_priv *rk817 = snd_soc_codec_get_drvdata(codec);
 
 	DBG("%s %d\n", __func__, mute);
+
 	if (mute) {
+		rk817_codec_ctl_gpio(rk817, CODEC_SET_SPK, 0);
+		rk817_codec_ctl_gpio(rk817, CODEC_SET_HP, 0);
+
 		snd_soc_update_bits(codec, RK817_CODEC_DDAC_MUTE_MIXCTL,
 				    DACMT_ENABLE, DACMT_ENABLE);
+		snd_soc_write(codec, RK817_CODEC_ADAC_CFG1,
+				PWD_DACBIAS_DOWN | PWD_DACD_DOWN |
+				PWD_DACL_DOWN | PWD_DACR_DOWN);
 		/* Reset DAC DTOP_DIGEN_CLKE for playback stopped */
 		snd_soc_update_bits(codec, RK817_CODEC_DTOP_DIGEN_CLKE,
 				    DAC_DIG_CLK_EN, DAC_DIG_CLK_DIS);
 		snd_soc_update_bits(codec, RK817_CODEC_DTOP_DIGEN_CLKE,
 				    DAC_DIG_CLK_EN, DAC_DIG_CLK_EN);
-	}
-	else {
+	} else {
 		snd_soc_update_bits(codec, RK817_CODEC_DDAC_MUTE_MIXCTL,
 				    DACMT_ENABLE, DACMT_DISABLE);
-	}
 
-	if (mute) {
-		rk817_codec_ctl_gpio(rk817, CODEC_SET_SPK, 0);
-		rk817_codec_ctl_gpio(rk817, CODEC_SET_HP, 0);
-	} else {
 		switch (rk817->playback_path) {
 		case SPK_PATH:
 		case RING_SPK:
